File: cpp/fast_modular_exponentation.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;

LL power(LL b, LL p, LL mod){
    if(p == 0)return 1;
    LL temp = power(b, p/2, mod);
    if(p % 2 == 0)
        return (temp * temp) % mod;
    else
        return (((b * temp) % mod) * temp) % mod;
}

File: python/fast_modular_exponentation.py
def mod_power(b, p, mod):
    if p == 0:
        return 1
    temp = mod_power(b, p//2, mod)
    if p % 2 == 0:
        return (temp * temp) % mod
    else:
        return (b * temp * temp) % mod
